synapse/http/replicationagent.py

```diff
@@ -0,0 +1,170 @@
+    copied from Twisted's Agent Class.
+
+    Attributes:
+        _endpointFactory: The IAgentEndpointFactory which will
+            be used to create endpoints for outgoing TCP connections.
+    """
+
+    def __init__(
+        self,
+        reactor: ISynapseReactor,
+        contextFactory: Optional[IPolicyForHTTPS] = None,
+        connectTimeout: Optional[float] = None,
+        bindAddress: Optional[bytes] = None,
+        pool: Optional[HTTPConnectionPool] = None,
+    ):
+        """
+        Create a ReplicationAgent.
+
+        Args:
+            reactor: A reactor for this Agent to place outgoing connections.
+            contextFactory: A factory for TLS contexts, to control the
+                verification parameters of OpenSSL.  The default is to use a
+                BrowserLikePolicyForHTTPS, so unless you have special
+                requirements you can leave this as-is.
+            connectTimeout: The amount of time that this Agent will wait
+                for the peer to accept a connection.
+            bindAddress: The local address for client sockets to bind to.
+            pool: An HTTPConnectionPool instance, or None, in which
+                case a non-persistent HTTPConnectionPool instance will be
+                created.
+        """
+        if not IPolicyForHTTPS.providedBy(contextFactory):
+            logger.warning(
+                f"{contextFactory} was passed as the HTTPS policy for an "
+                "Agent, but it does not provide IPolicyForHTTPS.  Since Twisted 14.0, "
+                "you must pass a provider of IPolicyForHTTPS.",
+            )
+            contextFactory = _DeprecatedToCurrentPolicyForHTTPS(contextFactory)
+
+        _AgentBase.__init__(self, reactor, pool)
+        endpoint_factory = ReplicationEndpointFactory(reactor, contextFactory)
+        self._endpointFactory = endpoint_factory
+
+    def _getEndpoint(self, uri: URI) -> IStreamClientEndpoint:
+        """
+        Get an endpoint for the given URI, using self._endpointFactory.
+            uri: The URI of the request.
+        Returns: An endpoint which can be used to connect to given address.
+        """
+        return self._endpointFactory.endpointForURI(uri)
+
+    def request(
+        self,
+        method: bytes,
+        uri: bytes,
+        headers: Optional[Headers] = None,
+        bodyProducer: Optional[IBodyProducer] = None,
+    ) -> "defer.Deferred[IResponse]":
+        """
+        Issue a request to the server indicated by the given uri.
+        An existing connection from the connection pool may be used or a new
+        one may be created.
+        Currently, HTTP and HTTPS schemes are supported in uri.
+
+        See: twisted.web.iweb.IAgent.request
+        """
+        # This function is overridden in preparation of future work:
+        # * So as to properly set a key for the pool and
+        # * to remove an _ensureValidURI() that will be in the way.
```

Comment:
```suggestion
        Issue a request to the server indicated by the given uri.

        An existing connection from the connection pool may be used or a new
        one may be created.

        Currently, HTTP and HTTPS schemes are supported in uri.

        This is copied from twisted.web.client.Agent, except:
        
        * It uses a different pool key (combining the host & port).
        * It does not call _ensureValidURI(...) since it breaks on some
          UNIX paths.

        See: twisted.web.iweb.IAgent.request
        """
```